File: language/benchmarks/tokens/algorithms/word-frequency/word-frequency.ts
type WordBucket = { count: number; word: string };

function frequency(lines: string[]): Map<string, number> {
  return normalizedWords(lines).reduce(increment, new Map<string, number>());
}

function increment(histogram: Map<string, number>, word: string): Map<string, number> {
  const nextHistogram = new Map(histogram);
  if (nextHistogram.has(word)) {
    nextHistogram.set(word, nextHistogram.get(word)! + 1);
  } else {
    nextHistogram.set(word, 1);
  }
  return nextHistogram;
}

function nonEmpty(word: string): boolean {
  return word !== "";
}

function normalizedWords(lines: string[]): string[] {
  return lines.flatMap(splitWords).filter(nonEmpty);
}

function splitWords(line: string): string[] {
  return line.toLowerCase().split(" ");
}

function summary(lines: string[]): WordBucket[] {
  return Array.from(frequency(lines).entries()).map(([key, value]) => ({ count: value, word: key }));
}

File: language/benchmarks/tokens/algorithms/word-frequency/word-frequency.py
WordBucket = dict[str, int | str]


def frequency(lines: list[str]) -> dict[str, int]:
    result: dict[str, int] = {}
    for word in normalized_words(lines):
        result = increment(result, word)
    return result


def increment(histogram: dict[str, int], word: str) -> dict[str, int]:
    next_histogram = dict(histogram)
    if word in next_histogram:
        next_histogram[word] = next_histogram[word] + 1
    else:
        next_histogram[word] = 1
    return next_histogram


def non_empty(word: str) -> bool:
    return word != ""


def normalized_words(lines: list[str]) -> list[str]:
    return [word for line in lines for word in split_words(line) if non_empty(word)]


def split_words(line: str) -> list[str]:
    return line.lower().split(" ")


def summary(lines: list[str]) -> list[WordBucket]:
    return [{"count": value, "word": key} for key, value in frequency(lines).items()]
